### PR TITLE
Speed up java version.

### PR DESCRIPTION
**Pull request by xupwup**
Redirecting output to /dev/null or gzip makes the program a lot faster.

### CODE REVIEW

File: java/WordCount.java [hunk 1]
```diff
@@ -1,48 +1,76 @@
...
-            Collections.sort(lst);
-            for (String word : lst) {
-                System.out.println(word + "\t" + count);
-            }
+        System.err.println("sorting...");
+        ArrayList<CountForWord> lst = new ArrayList<>(m.values());
+        Collections.sort(lst);
+        System.err.println("output...");
+        for(CountForWord c : lst){
+            System.out.println(c.word + "\t" + c.count);
```
Comment (svanoort) [inline java/WordCount.java]: Buffer your output: 

```java
// Now we print out the sorted words
        BufferedWriter log = new BufferedWriter(new OutputStreamWriter(System.out));

        for (CountForWord c : lst) {
            String outputStr = c.word + '\t' + c.count;
            log.write(outputStr, 0, outputStr.length());
            log.newLine();
        }
        log.flush();
        log.close();
```

**Comment by juditacs** [comment]
Could you please explain your changes in the code?

**Comment by xupwup** [comment]
I basically changed everything. 

The first change is to use a custom object in the hashmap, which avoids allocating lots of Integer objects (and also fixes the boxing/unboxing slowness). The second change is to remove the use of a regex to split the string, and instead code it manually. The third change is to avoid grouping by count, which saves a lot of processing.

**Comment by sgwerder** [comment]
Might wanna look and merge with my attempt
https://github.com/juditacs/wordcount/pull/29
WordCountArray.java
Anyway this is all in vain unless someone figures how to read a 3.6 GB InputStream quickly.

**Comment by xupwup** [comment]
> Might wanna look and merge with my attempt
> #29
> WordCountArray.java
> Anyway this is all in vain unless someone figures how to read a 3.6 GB InputStream quickly.

If I understand correctly you simply optimized the grouping part by replacing the tree by an array, correct? My solution does away with grouping altogether, so I don't see what your version has to offer.

**Comment by sgwerder** [comment]
> If I understand correctly you simply optimized the grouping part by replacing the tree by an array, correct? My solution does away with grouping altogether, so I don't see what your version has to offer.

It inserts into the primitive Array at the position of count of each String. Therefor it gets by without any need for sorting or comparison on insert. A lot of conditionals can be safed that way and also Boxing / Unboxing / Creating Objects doesnt need to happen.
Drawback is having a sparsly filled Array, but who cares for that?

**Comment by juditacs** [comment]
Can you guys work out a common solution or do you want me to compare this PR and https://github.com/juditacs/wordcount/pull/29 ?

**Comment by sgwerder** [comment]
I guess working out a common solution? As far as my tests show for wikisource:
- WordCount Original:  7.67 s
- WordCount XupWup: 7.03 s
- WordCountArray:      6.97 s
- WordCountFastUtil:   6.53 s

File: java/WordCount.java [hunk 2]
```diff
@@ -1,48 +1,76 @@
...
+        }
+
+        @Override
+        public int compareTo(CountForWord t) {
+            if(count < t.count){
+                return 1;
+            }else if(count > t.count){
+                return -1;
+            }else{
+                return -word.compareTo(t.word);
```
Comment (sgwerder) [inline java/WordCount.java]: Do not negate here to sort lexicographically and make tests pass.

**Comment by xupwup** [comment]
Try comparing his solution with mine, I strongly doubt that his idea will help much. It seems like a blend between counting sort for the counts and normal sort for the strings, but I suspect that it won't help.

File: java/WordCount.java [hunk 3]
```diff
@@ -69,8 +71,13 @@ public static void main(String[] args) throws IOException {
         ArrayList<CountForWord> lst = new ArrayList<>(m.values());
         Collections.sort(lst);
         System.err.println("output...");
+        BufferedWriter outputWriter = new BufferedWriter(new OutputStreamWriter(System.out));
         for(CountForWord c : lst){
-            System.out.println(c.word + "\t" + c.count);
+            outputWriter.write(c.word);
```
Comment (sgwerder) [inline java/WordCount.java]: `outputWriter.write(c.word + "\t" + c.count);
 outputWriter.newLine();`

Otherwise it's going to interpret c.count as a char

**Comment by xupwup** [comment]
I've merged @sgwerder's solution to a separate file. Let's see if it speeds things up. (I haven't tested it)

**Comment by juditacs** [comment]
Please add them to run_commands.txt and build.sh.

**Comment by sgwerder** [comment]
Wait a second, I've been adopting your read and write method.

**Comment by sgwerder** [comment]
Specifing a huge buffer size bigger than default (8192) gives it yet another boost on my end:
`BufferedReader br = new BufferedReader(new InputStreamReader(System.in), 1048576);`

**Comment by svanoort** [comment]
@xupwup  This is probably near optimal in all but I/O.  

The approach here that uses the Count object (and avoids boxing/unboxing) is probably as close as you are going to get to optimal for the actual count/sort without writing a custom HashMap implementation that allows you to directly expose the backing arrays for iteration. 

The custom parsing is as CPU efficient as you're likely to see using string objects. 

At this point I think all optimization probably needs to be on the I/O front: readLine is somewhat inefficient and requires creating throwaway string objects here, with all the implied copies.  Additionally, BufferedEeaders have more overhead than directly reading a block at a time to arrays. 

We are likely to see improved performance by reading large blocks  to a char[] directly from an InputStreamReader wrapping the inputstream.  Then the parser can add '\n' and '\r' as separator tokens and directly iterate on that array. The only tricky part here is when word boundaries cross our buffer array boundaries.  The easiest thing to do is probably to carry the first/last bits to a StringBuilder. 

I've seen significant speedups with this approach for dense binary data.  It's possible that use of NIO would offer some small benefits too.

**Comment by sgwerder** [comment]
@svanoort It is indeed choking with the IO atm. It can't read the 3.6 GB test file and so doesnt even get to process anything.

**Comment by xupwup** [comment]
Is the test file only one line or something?

**Comment by sgwerder** [comment]
The test file: http://dumps.wikimedia.org/huwiki/latest/huwiki-latest-pages-meta-current.xml.bz2

**Comment by xupwup** [comment]
The changes to NIO don't seem to help much on my end though...

**Comment by sgwerder** [comment]
Huwikisource doesn't pass test anymore and run time is slightly slower. The whole huwiki still doesnt read in either.

**Comment by xupwup** [comment]
Why does the whole wiki source not work, though? Does the hashmap grow too big or something?

**Comment by sgwerder** [comment]
> Why does the whole wiki source not work, though? Does the hashmap grow too big or something?
View changes
I don't know for this one. All other tests pass fine. Maybe has to do with the size. Here's the wikisource (around 120MB): http://dumps.wikimedia.org/huwikisource/latest/huwikisource-latest-pages-meta-current.xml.bz2
You can create the solution with the original version. Here's the simple thing I used to test:

`package wordcount;

import static org.junit.Assert.assertEquals;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintStream;
import java.io.PrintWriter;
import java.util.Scanner;

public class WordCountRun {

	private static PrintStream STDOUT = System.out;
	private static ByteArrayOutputStream OUT_CONTENT = new ByteArrayOutputStream();
	private static String FILE_NAME = "huwiki";

	public static void main(String[] args) throws IOException {
		reset();
		long start = System.nanoTime();
		WordCountMergedNIO.main(new String[0]);
		long end = System.nanoTime();
		long time = end - start;
		// assertEquals(readFile("./data/test/" + FILE_NAME + ".out"), OUT_CONTENT.toString());
		// writeFile("./data/test/" + FILE_NAME + ".out");
		print("WordCount", time);
		reset();
	}

	private static void reset() {
		OUT_CONTENT = new ByteArrayOutputStream();
		try {
			System.setIn(new FileInputStream("./data/test/" + FILE_NAME + ".in"));
		}
		catch (FileNotFoundException e) {
			e.printStackTrace();
		}
		System.setOut(new PrintStream(OUT_CONTENT));
	}

	private static void print(String name, long time) {
		System.setOut(STDOUT);
		System.out.println(name + ": " + time / 1000000000.0);
		System.setOut(new PrintStream(OUT_CONTENT));
	}

	private static String readFile(String pathname) throws IOException {
		File file = new File(pathname);
		StringBuilder fileContents = new StringBuilder((int) file.length());
		Scanner scanner = new Scanner(file);
		String lineSeparator = System.getProperty("line.separator");

		try {
			while (scanner.hasNextLine()) {
				fileContents.append(scanner.nextLine() + lineSeparator);
			}
			return fileContents.toString();
		}
		finally {
			scanner.close();
		}
	}

	private static void writeFile(String pathname) {
		File f = new File(pathname);
		PrintWriter out;
		try {
			out = new PrintWriter(f);
			out.write(OUT_CONTENT.toString());
			out.close();
		}
		catch (FileNotFoundException e) {
			e.printStackTrace();
		}
	}

}`

**Comment by xupwup** [comment]
That code keeps all output on stdout in memory until the program finishes. You probably don't want that.

**Comment by svanoort** [comment]
Well, I tried a bulk-read-and-parse implementation here, which doesn't tell the inputstream to do any readLine parsing: https://github.com/svanoort/wordcount/tree/svanoort-fast-java

It reads blocks of configurable size, then digests them as a chunk. What I think is interesting is that somehow Java manages to throw an OutOfMemory error on this, even though it's only doing streamwise reads.  Possibly due to the large amount of object creation during tokenizing?  I do see that GC time runs at 10-15%, with a 2 GB heap (-Xmx2048m).  

Eventually it hits 'Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded' which says that either the working set (the HashMap and its count objects, plus strings, plus the buffer array) is growing excessively large *or* the sheer amount of string garbage generated is eventually choking GC.

**Comment by svanoort** [comment]
Ugh, 1.2M unique strings on the wikisource alone.  Probably the amount of memory overhead for the default hashtable + string overheads + object overheads in Java are a problem with the full encyclopedia, since we're only splitting on whitespace (not on punctuation), and it's a case-sensitive count of uniques. 

Full wiki output reports 23494121 unique entries, yes 23.5 MILLION.   

Doing a count of actual characters in the strings alone:
cat data/huwiki-latest-pages-meta-current.xml | sed 's/[\t ]/\n/g' | grep -v ^$ | sort | uniq | wc -m

At this point, I think the only solution to fitting it all in reasonable memory in Java is a trove [TObjectIntCustomHashMap](http://trove4j.sourceforge.net/javadocs/gnu/trove/map/custom_hash/TObjectIntCustomHashMap.html#TObjectIntCustomHashMap()) that maps byte[]  (raw Unicode bytes) to int.   The hashing strategy would be to convert the bytes[] (which are the UTF-8 decoded string contents from input) to a string and then do a hashcode. For the equals method, it would compare array contents.  That has about as little overhead per entry as possible, 8 bytes for the key pointer and int values in backing HashMap backing arrays (plus 25% empty hashmap slots), 12 bytes per key entry (for the array object and its ize) + 1 byte per character in UTF-8 encoding.    

Call it 20-24B per entry + raw text content (not fluffed up).  Then there's padding of objects in memory for the strings. 

Beats the String + its fields + char[] (2 bytes per character in memory) + Hashmap Entry object + LinkedList of entries for each slot in the hashmap array + all their overheads.

**Comment by sgwerder** [comment]
Can you implement that? This might help, check it out before: http://java-performance.info/hashmap-overview-jdk-fastutil-goldman-sachs-hppc-koloboke-trove-january-2015/

**Comment by svanoort** [comment]
@sgwerder Yes, I could probably implement it but I think the benchmark needs a different benchmark dataset; this one has too many unique "words" in it relative to total filesize, and is no longer a good language benchmark at all.  If 2.5 GB of 4.3 GB is unique, you're not really counting anymore.

I've seen the large datastructures benchmark; really, the reason to prefer Trove here is the API and simplicity (it should require less classes), and that it allows you to directly apply an operation on inputs.

This specific method is perfect for our use case: http://trove4j.sourceforge.net/javadocs/gnu/trove/map/hash/TObjectIntHashMap.html#adjustOrPutValue(K, int, int)

Though I do see the FastUtil option is rather nice too and has similar interfaces: http://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/objects/Object2IntOpenCustomHashMap.html

Really the full-wiki test is designed to destroy languages that use UTF-16 internally to store unicode strings (as opposed to UTF-8 or worse yet converting to bytes).  Java just gets hammered all around between this, lack of native primitive collections, object memory overheads, and the fact that strings are immutable.

And the bash script 'cheats' because the sort command buffers results to disk as it processes (using ~4 GB of temp disk).

**Comment by sgwerder** [comment]
@juditacs whats your opinion ?

**Comment by xupwup** [comment]
Maybe we could use a [Trie](https://en.wikipedia.org/wiki/Trie) instead of a hashmap. That would avoid having to store every string in its entirety. In the example on wikipedia they let the roots of the tree contain the full string, but you could remove the parts that are defined in the parent nodes to save space.

**Comment by sgwerder** [comment]
I encountered your Trie suggestion too, where I also got the idea with the array insertion: http://cs.stackexchange.com/a/26434
I'm sceptical about Trie's though. Might save memory but could use too much time building and traversing it.

**Comment by xupwup** [comment]
It should be about as fast as a hashmap, I guess.

**Comment by juditacs** [comment]
Python3 stores strings using UTF-16 internally, this shouldn't be a problem.

You're welcome to test with a different corpus on your local machine but the distribution of the words won't change significantly.

**Comment by juditacs** [comment]
I want to remove one of the java solutions and keep one WordCount.java. Which one should stay?

**Comment by sgwerder** [comment]
WordCountArray I'd say.

**Comment by juditacs** [comment]
Tokenization is a lot harder than stripping punctuation and there are complete libraries for this. Reimplementing a sophisticated tokenization algorithm in each language would be an interesting task.

WordCount is slightly faster than WordCountEntries. Are you sure about keeping WordCountEntries?

**Comment by sgwerder** [comment]
Ahh no, I thought you meant only one file in total. Then keep WordCount and WordCountArray.

**Comment by juditacs** [comment]
I did mean only one file.

Right now WordCount is faster than WordCountArray, but you told me to keep WordCountEntries.

**Comment by sgwerder** [comment]
Keep the fastest.

**Comment by svanoort** [comment]
If this is merged, I'll offer up the Trove version as a separate PR.  I've got an initial version done, just working through some classpath issues (*sigh*).  

v1 still uses String objects (overhead, I know) and sortable CountForWord objects but should be faster on parse/count and have less hashTable overheads per entry.  It may hit memory issues when it has to create the CountForWord objects to sort them though.  

Then I can tinker with some clever bit-twiddling to allow sorting without creating CountForWord objects (which carry a 12-byte overhead minimum, probably 16 with 8-byte memory alignment), and using char[] in place of String (savings: 12-16 bytes per string).   Savings overall: probably about 1 GB of RAM, peak of 2 GB when copying from HashMap to list for sorting. 

Packing char\[ \] to UTF-8 byte\[ \] will be the other option for saving RAM, but I think will come with a gigantic performance overhead.

**Comment by sgwerder** [comment]
> Packing char[ ] to UTF-8 byte[ ] will be the other option for saving RAM, but I think will come with a gigantic performance overhead.

I think you could read in UTF-8 byte[ ] instead of char[] directly or do you mean something else?

**Comment by svanoort** [comment]
@sgwerder For correct handling, it needs to parse the bytes into Unicode (before tokenizing) to read, but for the map representation you can convert to UTF-8 byte\[ \] objects to store in memory.  This is exactly what the Python 3 version does.  (Although I think the Py2 versions are not handling Unicode correctly, I think they will not handle multibyte characters right and  may sbut it's tricky to create a testcase for this because you need multibyte Unicode characters with newline/space characters in their bytes).  

I like JEP-254, it will make a huge difference in performance & memory!  I kind of wish they were just using UTF-8 everywhere, but I'm assuming the added processing complexity there offsets memory gains here.  Would also be nice to see more of the APIs support use of StringBuilders rather than constantly creating String objects all over the place too (one aspect of Java that is a major performance killer).  So many bad decisions with String handling in Java *sigh.*  It's easier to make things 'correct' but everything is slowed down because it's enforced where not needed.

By the way: @juditacs it's worth mentioning that the Python implementations here are doing this same trick I'm proposing for Java with Unicode -> raw UTF-8 bytes as dictionary keys.  The only time you see full Unicode strings in memory is when they're read in or written out, the map and sorting is done on byte strings.  I did a quick test, and with the Py3 version modified to store full unicode in memory you go up to to 6.8+ GB of memory (full UTF-16, Unicode objects).  In comparison, Py2gabor runs in 2.8 GB.  There's a catch here too for the Python 2 versions: AFAICT they are never explicitly converting to Unicode before parsing to lines or tokens, so it's quite possible that multibyte Unicode characters containing the space/newline characters will show (incorrectly) as spaces.

The fact that  Python APIs are friendly to using raw bytes objects in maps is something of a Py2 legacy (one small advantage of that painful transition).  Super convenient here though.

**Comment by juditacs** [comment]
IMHO, in UTF-8 only the 0-127 range can have 0 as the MSB, so no confusion is possible.

**Comment by xupwup** [comment]
In that case you don't actually have to parse unicode, right? You can just look for bytes with value ' ', '\n' and '\t', right? [This](https://github.com/juditacs/wordcount/pull/27/commits/ff831eda57224ef4ac834f2ac631844f4ada99b0) broken commit did that, but still ran out of memory.

**Comment by svanoort** [comment]
@juditacs You are correct, I had forgotten that the high bit is always 1 for successive bytes, so collisions are not a point of concern.

@xupwup / @sgwerder Trove-enabled version using byte arrays instead of strings here https://github.com/svanoort/wordcount/tree/svanoort-trove - it's about 20% faster on my machine, and will complete the full wiki with about 4G of heap (I allocate 6G in the run command to give some extra headroom and limit GC pressure).  

I should probably clean that version up into a separate commit though.

**Comment by juditacs** [comment]
Yes, unicode parsing is not necessary at all. It would be very interesting to see how well unicode is supported by each language, maybe next time.

**Comment by sgwerder** [comment]
@svanoort Yay, it completed. 257 seconds. While I had a lot of background stuff running. Unforunately it isn't passing the little tests. Comparing the outputs, I didn't see the fault immediately. Maybe a \n missing at the end?
Sorting might be done in parallel (which fastutil offers)

**Pull request merged**